### PR TITLE
Implement linalg lowering of diag_embed torch op

### PR DESCRIPTION
**Pull request by schnkmwt**
This PR adds lowering of diag_embed to linalg dilect.
Tracked in https://github.com/nod-ai/SHARK-Turbine/issues/288

### CODE REVIEW

File: lib/Conversion/Utils/Utils.cpp [hunk 1]
```diff
@@ -166,6 +166,33 @@ castIndexVectorToInt64Vector(OpBuilder &b, Location loc,
...
 
+SmallVector<Value> getDiagEmbedResultShape(OpBuilder &b, Location loc, Value tensor, int64_t  offset, int64_t  dim1, int64_t  dim2) {
+  auto inputType = tensor.getType().cast<RankedTensorType>();
+  auto inputRank = inputType.getRank();
+  auto resultRank = inputRank + 1;
+  
+  SmallVector<Value> resultShape;
+  Value constZero = b.create<arith::ConstantIndexOp>(loc, 0);
+  Value constNegOne = b.create<arith::ConstantIndexOp>(loc, -1);
+  Value constOffset = b.create<arith::ConstantIndexOp>(loc, offset);
```
Comment (ramiro050) [inline lib/Conversion/Utils/Utils.cpp:177]: nit: offset is a `Value` originally, so you can just pass that `Value` as the `offset` argument to `getDiagEmbedResultShape`

File: lib/Conversion/Utils/Utils.cpp [hunk 2]
```diff
@@ -166,6 +166,33 @@ castIndexVectorToInt64Vector(OpBuilder &b, Location loc,
...
+  Value isNegOffset = b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::slt, constOffset, constZero);
+  Value mulOffsetNegOne = b.create<arith::MulIOp>(loc, constOffset, constNegOne);
+  Value absOffset = b.create<arith::SelectOp>(loc, isNegOffset, mulOffsetNegOne, constOffset);
```
Comment (ramiro050) [inline lib/Conversion/Utils/Utils.cpp:180]: nit: I think you can use `math::AbsIOp`

**Review by ramiro050 — CHANGES_REQUESTED** [review]
Thanks! Looks mostly good, I just have a few comments

File: projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py [hunk 3]
```diff
@@ -53,6 +53,33 @@ def _embedding_bag_helper(weight: List[int], indices: List[int],
 
     return output_bag_shape, offset2bag_shape, bag_size_shape, max_indices_shape
 
+# TODO: upstream this
```
[the remark this answers is not in the document]
Comment (schnkmwt) [reply projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py]: Yes I forgot to add those. Added

File: lib/Conversion/Utils/Utils.cpp (hunk 1 above)
Comment (schnkmwt) [reply lib/Conversion/Utils/Utils.cpp]: I think I wrote it separately first so cleaned it up now.

File: lib/Conversion/Utils/Utils.cpp (hunk 2 above)
Comment (schnkmwt) [reply lib/Conversion/Utils/Utils.cpp]: Good suggestion! I'm not very well versed in all available ops and dilects yet. Replaced with `math::AbsIOp`

File: lib/Conversion/TorchToLinalg/DataMovement.cpp [hunk 4]
```diff
@@ -2000,6 +2001,157 @@ class ConvertAtenDiagonalOp : public OpConversionPattern<AtenDiagonalOp> {
...
+    if (!matchPattern(op.getOffset(), m_TorchConstantInt(&offset)))
+      return rewriter.notifyMatchFailure(op, "offset is not constant");
+
+    int64_t dim1;
+    if (!matchPattern(op.getDim1(), m_TorchConstantInt(&dim1)))
+      return rewriter.notifyMatchFailure(op, "dim1 is not constant");
+    dim1 = toPositiveDim(dim1, resultRank);
+    if (!isValidDim(dim1, resultRank))
+      return rewriter.notifyMatchFailure(
+          op, "dim1 can only be between [" + std::to_string(-resultRank) + "," +
```
Comment (newling) [inline lib/Conversion/TorchToLinalg/DataMovement.cpp:2061]: nit:
```suggestion
          op, "dim1 can only be in closed range [" + std::to_string(-resultRank) + "," +
```

File: lib/Conversion/TorchToLinalg/DataMovement.cpp [hunk 5]
```diff
@@ -2000,6 +2001,157 @@ class ConvertAtenDiagonalOp : public OpConversionPattern<AtenDiagonalOp> {
...
+    if (!matchPattern(op.getDim2(), m_TorchConstantInt(&dim2)))
+      return rewriter.notifyMatchFailure(op, "dim2 is not constant");
+    dim2 = toPositiveDim(dim2, resultRank);
+    if (!isValidDim(dim2, resultRank))
+      return rewriter.notifyMatchFailure(
+          op, "dim2 can only be between [" + std::to_string(-resultRank) + "," +
+                  std::to_string(resultRank - 1) + "]");
+
+    if (dim1 == dim2)
+      return rewriter.notifyMatchFailure(op, "dim1 and dim2 can not be equal");
```
Comment (newling) [inline lib/Conversion/TorchToLinalg/DataMovement.cpp:2074]: I think this particular case would be better handled in the op verifier (ops can be made to have verifiers in `pt1/python/torch_mlir/jit_ir_importer/build_tools/torch_ods_gen.py` -- see for example aten::permute). 

ps definitely not a required change as almost no ops have verifiers at the moment.

**Review by newling — COMMENTED** [review]
I'm still trying to understand how the indexing works. Test coverage looks good, and the comments are useful. Nice one!

**Comment by ramiro050** [comment]
Not sure if you're manually making the lint fixes, but you can use `git clang-format` to automatically format the changes.

```
git clang-format main
```

formats your entire branch. If you don't have `git clang-format`, you should be able to do inside `torch-mlir/`

```
./externals/llvm-project/clang/tools/clang-format/git-clang-format main

File: lib/Conversion/TorchToLinalg/DataMovement.cpp (hunk 4 above)
Comment (schnkmwt) [reply lib/Conversion/TorchToLinalg/DataMovement.cpp]: Thanks Done!

File: lib/Conversion/TorchToLinalg/DataMovement.cpp (hunk 5 above)
Comment (schnkmwt) [reply lib/Conversion/TorchToLinalg/DataMovement.cpp:2169]: Thanks James, I'll keep it mind for future.

**Comment by schnkmwt** [comment]
> Not sure if you're manually making the lint fixes, but you can use `git clang-format` to automatically format the changes.
> 
> ```
> git clang-format main
> ```
> 
> formats your entire branch. If you don't have `git clang-format`, you should be able to do inside `torch-mlir/`
> 
> ```
> ./externals/llvm-project/clang/tools/clang-format/git-clang-format main
> ```

For some reason `git clang-format main` did not work for me. I guess it only works on most recent commit. But the `git-clang-format` under `./externals` worked like a charm! Thanks for the suggestion.

**Review by ramiro050 — APPROVED** [review]
LGTM, I think you need to XFAIL your tests on the ONNX backend

**Comment by schnkmwt** [comment]
> LGTM, I think you need to XFAIL your tests on the ONNX backend

Yes just pushed that. Got help from discord!

**Review by newling — APPROVED** [review]

**Pull request merged**